### PR TITLE
Ordered compiler names

### LINKED ISSUE #518: Auto-detection of compilers with multiple names
Recently, I discovered that certain PGI compilers (`pgf77` and `pgf90`) were unable to build HDF4, while newer ones (`pgfortran`) had no trouble. I submitted PR #493 to solve this issue. Now, if your compilers.yaml contains something like:

``` yaml
pgi@15.10-0:
      cc:  /soft/pgi/15.10/linux86-64/15.10/bin/pgcc
      cxx: /soft/pgi/15.10/linux86-64/15.10/bin/pgc++
      f77: /soft/pgi/15.10/linux86-64/15.10/bin/pgfortran
      fc:  /soft/pgi/15.10/linux86-64/15.10/bin/pgfortran
```

HDF4 will build for you. However, when I run:

``` bash
spack compiler add /soft/pgi/15.10/linux86-64/15.10/bin
```

Spack still finds the older compilers first:

``` yaml
pgi@15.10-0:
      cc:  /soft/pgi/15.10/linux86-64/15.10/bin/pgcc
      cxx: /soft/pgi/15.10/linux86-64/15.10/bin/pgCC
      f77: /soft/pgi/15.10/linux86-64/15.10/bin/pgf77
      fc:  /soft/pgi/15.10/linux86-64/15.10/bin/pgf90
```

The [PGI Installation Guide](https://www.pgroup.com/doc/pgiinstall.pdf) recommends using the newer compilers, although the older ones are kept around for backwards compatibility. Although I could simply remove `pgCC`, `pgf77`, and `pgf90` from those lists, it would cause problems for anyone using older installations of PGI, which don't come with the newer compilers.

I propose we change the way that `spack compiler add` finds compilers. My suggestion would be to choose the first match it finds in `cxx_names`, `f77_names`, and `f90_names`. I believe the relevant code is in the `find` method of `lib/spack/spack/compiler.py`. Things start to get complicated since we also need to be able to find compilers with prefixes, suffixes, or version numbers, and much of the code is parallelized for efficiency.

If anyone would like to tackle this, and maybe wrap in a solution to any of the unresolved problems we've been seeing with Apple compilers, be my guest. Otherwise, I will hopefully get to it eventually. It doesn't significantly impact me since I can just `spack config edit compilers` and switch it over to the newer ones, but it might trip up a new Spack user.

### PR DESCRIPTION
**Pull request by becker33**
Fixes #518 

Summary:
Previously, when there were multiple names for a compiler, Spack chose the name that occurs first lexicographically as it searches the directory. This caused problems for the PGI compiler, for which we should choose `pgfortran` when available over `pgf77` and `pgf90`.

@adamjstewart proposed in #518 that a reasonable solution would be to choose compilers in the order they appear in the compiler_names list in Spack. This allows us to set priorities for compiler names using the compiler subclasses in Spack.

This PR implements that solution, and adds a test for it. Also adds a test to ensure Spack chooses the first compiler of a given version in the paths provided to `spack compiler add`.

### CODE REVIEW

**Comment by adamjstewart** [comment]
@becker33 can you rebase?

**Comment by becker33** [comment]
@adamjstewart sorry missed your mention before, it's rebased now.

File: lib/spack/spack/compiler.py [hunk 1]
```diff
@@ -261,15 +262,21 @@ def _find_matches_in_path(cls, compiler_names, detect_version, *path):
...
         for directory in path:
+            dir_checks = []
             if not (os.path.isdir(directory) and
                     os.access(directory, os.R_OK | os.X_OK)):
                 continue
 
             files = os.listdir(directory)
             for exe in files:
-                full_path = os.path.join(directory, exe)
+                full_path = join_path(directory, exe)
```
Comment (adamjstewart) [inline lib/spack/spack/compiler.py]: I prefer using `os.path.join` internally, `join_path` is more of a shorthand designed for packages. You already need to `import os` anyway, so this just adds a second unnecessary import.

File: lib/spack/spack/test/cmd/test_compiler_cmd.py [hunk 2]
```diff
@@ -63,6 +66,83 @@ def mock_compiler_dir(tmpdir):
     return str(tmpdir)
 
 
+@pytest.fixture()
+def mock_two_compiler_dirs(tmpdir):
+    """Return two directory containing a fake, but detectable compiler,
```
Comment (adamjstewart) [inline lib/spack/spack/test/cmd/test_compiler_cmd.py]: directory -> directories

File: lib/spack/spack/compiler.py (hunk 1 above)
Comment (becker33) [reply lib/spack/spack/compiler.py]: Yeah that's a rebase artifact, I'll fix that.

File: lib/spack/spack/test/cmd/test_compiler_cmd.py (hunk 2 above)
Comment (becker33) [reply lib/spack/spack/test/cmd/test_compiler_cmd.py]: Fixed

**Comment by adamjstewart** [comment]
I figured out how to reproduce #518. If I change the order of the compilers in `lib/spack/spack/compilers/pgi.py`, `develop` still gives me the same binaries, but your branch gives me the binary that comes first in the list. So now I can confirm that this works. Should be safe to merge once the tests pass.

**Comment by adamjstewart** [comment]
The doc tests have stalled 3 times in a row now. Not sure what else to try.

**Comment by tgamblin** [comment]
I rebased on `develop`, and the tests still fail and get stuck in docs.  When I run `spack compiler find` on my mac on this branch, it fails.

I think the doc test issues are really Spack issues that are manifesting in things that the docs call.  Specifically, I think something is failing in `spack compiler find`, which will be run when Spack is first run in the doc environment.  

I added a commit that moves the doc tests to be parallel with the unit tests, which may explain more.  If the unit test succeed now, there is probably something we're not checking thoroughly enough, and @becker33 should add a test for it.

File: lib/spack/spack/compiler.py [hunk 3]
```diff
@@ -261,8 +261,14 @@ def _find_matches_in_path(cls, compiler_names, detect_version, *path):
         prefixes = [''] + cls.prefixes
         suffixes = [''] + cls.suffixes
 
+        def check_cmp_key(check):
+            name = os.path.basename(check[0])
+            idx = compiler_names.index(name)
```
Comment (tgamblin) [inline lib/spack/spack/compiler.py]: This fails when `check[0]` is not in the list, and it causes `spack compiler find` to stall.

Comment (becker33) [reply lib/spack/spack/compiler.py]: @tgamblin I think I introduced this bug when I rebased. Fixed it now.

**Comment by becker33** [comment]
@tgamblin can you confirm that I've addressed your changes?

**Comment by adamjstewart** [comment]
Ping @tgamblin, needs a rebase @becker33

**Comment by becker33** [comment]
ping @tgamblin

File: lib/spack/spack/compiler.py [hunk 4]
```diff
@@ -263,12 +267,17 @@ def _find_matches_in_path(cls, compiler_names, detect_version, *path):
...
 
                     match = re.match(regex, exe)
                     if match:
                         key = (full_path,) + match.groups() + (detect_version,)
-                        checks.append(key)
+                        dir_checks.append(key)
+
+            # sort dir_checks by compiler name order
+            # this allows us to prioritize compiler names in subclass
+            dir_checks = sorted(dir_checks, key=check_cmp_key)
```
Comment (tgamblin) [inline lib/spack/spack/compiler.py]: I'd either move the check_cmp_key definition closer to this `sorted()` or just use a `lambda c: compiler_names.index(c[2])` on a new line.  I found it hard to read this currently as what a "check" is doesn't become apparent until you read the loop.

Comment (becker33) [reply lib/spack/spack/compiler.py]: @tgamblin refactored.

**Comment by adamjstewart** [comment]
Ping @tgamblin 

@becker33 can you rebase?

**Comment by becker33** [comment]
@adamjstewart I attempted to rebase this, but the intervening changes made it easier to start from scratch and copy over the new tests. Closing this in favor of #17449

**Pull request closed without merging**